More talking

ekwhite opened the pull request:
talk pull

Commit:
added Bind, Strafter (buggy), changed Filter

Commit:
added more sparql keywords and explicit filter statements

sinistral commented on kr-core/src/main/clojure/edu/ucdenver/ccp/kr/sparql.clj:
```diff
@@ -243,11 +244,43 @@
+
+
+
+
```
Does the specificity of the method name add enough value to justify this duplication?  I feel as though `(operator-body "IRI" [clauses])` isn't significantly less readable than `(iri-body [clauses])`

Alternatively, we could create a macro that would enable something like `(defoperator iri)` to automatically define the appropriate function without the duplication.

sinistral commented on kr-core/src/main/clojure/edu/ucdenver/ccp/kr/sparql.clj:
```diff
@@ -243,11 +244,43 @@
+(defn naked-unary-operator [op-name]
+  (fn [args]
+    (str " " op-name (operator-body (first args)))))
+
```
`naked-unary-operator` and `unary-operator` could be collapsed into a single function that takes a keyword argument `:naked?` (which defaults to `false`) to indicate whether the parens are needed.  

Either way, it would be nice to have some documentation on the function that describes when each form is appropriate.

sinistral commented on kr-core/src/main/clojure/edu/ucdenver/ccp/kr/sparql.clj:
```diff
@@ -304,6 +337,12 @@
       :str (unary-operator "str")
       :lang (unary-operator "lang")
       :datatype (unary-operator "datatype")
+      ;; duplicate bind
```
What does this comment mean?

sinistral commented on kr-core/src/main/clojure/edu/ucdenver/ccp/kr/sparql.clj:
```diff
@@ -404,7 +439,14 @@
                             ;;(repeat ". \n")))
                             (repeat " \n")))
-   (sparql-operator? (first triple-pattern)) (filter-body triple-pattern)
+   ;; used as a catch-all; this routes anything unusual to a FILTER.
```
I found this comment confusing, since (I think, correct me if I'm wrong) the _this_ that it refers to is the line that is commented out.  Since that functionality has been superseded, I'd suggest just removing both code and comment.  If you put a good explanation in the commit log, that will make it clear in future why this changes was necessary and how it differs from the old behaviour.

sinistral commented on kr-core/src/main/clojure/edu/ucdenver/ccp/kr/sparql.clj:
```diff
@@ -430,7 +474,11 @@
+    (println "options" options)
+    (println "non-vars" non-vars)
+    (println "prefixes" prefixes)
+    (println "namespaces" namespaces)
```
Are these `println` statements will necessary, or were they added for debugging?  If they're generally useful, we should start using [logging](https://github.com/clojure/tools.logging) to report these so that we have more control over the output.

sinistral commented on kr-core/src/main/clojure/edu/ucdenver/ccp/kr/sparql.clj:
```diff
@@ -243,11 +244,43 @@
+       (operator-body concat-clauses)
+       " \n "))
+
+(defn strafter-body [strafter-clauses]
```
Your summary indicates that `strafter` is buggy; can you add more detail to the commit log about what's missing/broken?

sinistral commented:
I've made a few comments in-line.  This seems like a quite a nice pragmatic change, given what we have to work with.  

More generally: would it be possible to add tests cases for the new operators, please?  I know that we want to move away from build compliant SPARQL ourselves, but even when happens, we're unlikely to change all of the existing rules, and it'll be important to ensure that these operators continue to behave correctly.  The test cases will also also serve as demonstrations of the functionality, which makes it easier to see what they do without having to dig through the rules.

sinistral commented on kr-core/src/main/clojure/edu/ucdenver/ccp/kr/sparql.clj:
```diff
@@ -404,14 +466,23 @@
                             (repeat " \n")))
-   (sparql-operator? (first triple-pattern)) (filter-body triple-pattern)
+   ;; used as a catch-all; this routes anything unusual to a FILTER.
+   ;; (sparql-operator? (first triple-pattern)) (filter-body triple-pattern)
```
This change is going to break backwards compatibility.  Before operators were implicitly wrapped in a filter, whereas now the filter body must be explicitly specified.

I really should have seen this earlier  ):

ekwhite replied:
True and not surprising.  What is the solution, though?  Should we maintain 2 versions, and consider the new one as a sparql 1.1 version?  I remember worrying about this issue, but am not sure how to deal with sparql 1.1 syntax when everything is wearing a filter.  Maybe tomorrow we can talk more about this.

From: Marc Daya notifications@github.com
Reply-To: UCDenver-ccp/kr reply@reply.github.com
Date: Tuesday, May 31, 2016 at 10:50 AM
To: UCDenver-ccp/kr kr@noreply.github.com
Cc: Elizabeth White elizabeth.white@raritanfarm.com, Author author@noreply.github.com
Subject: Re: [UCDenver-ccp/kr] More talking (#4)

In kr-core/src/main/clojure/edu/ucdenver/ccp/kr/sparql.clj:

> @@ -404,14 +466,23 @@
>        (apply str (interleave (map sparql-query-body triple-pattern)
>                               ;;(repeat ". \n")))
>                               (repeat " \n")))
> -   (sparql-operator? (first triple-pattern)) (filter-body triple-pattern)
> +   ;; used as a catch-all; this routes anything unusual to a FILTER.
> +   ;; (sparql-operator? (first triple-pattern)) (filter-body triple-pattern)
> This change is going to break backwards compatibility. Before operators were implicitly wrapped in a filter, whereas now the filter body must be explicitly specified.

I really should have seen this earlier ):

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

sinistral commented on kr-core/src/main/clojure/edu/ucdenver/ccp/kr/sparql.clj:
```diff
@@ -21,7 +21,7 @@
 ;;; dynamic variables to control function
 
-(defonce ^:dynamic *select-type* select-default)
+(defonce ^:dynamic *select-type* select-distinct)
```
What is the motivation for changing the default?

ekwhite replied:
Whoops; I think it’s not needed.  That’s a relic from when I was trying to write ice rules, and is not used.  Sorry!

From: Marc Daya notifications@github.com
Reply-To: UCDenver-ccp/kr reply@reply.github.com
Date: Tuesday, May 31, 2016 at 3:27 PM
To: UCDenver-ccp/kr kr@noreply.github.com
Cc: Elizabeth White elizabeth.white@raritanfarm.com, Author author@noreply.github.com
Subject: Re: [UCDenver-ccp/kr] More talking (#4)

In kr-core/src/main/clojure/edu/ucdenver/ccp/kr/sparql.clj:

> @@ -21,7 +21,7 @@
>   
>   ;;; dynamic variables to control function
>   
> -(defonce ^:dynamic _select-type_ select-default)
> +(defonce ^:dynamic _select-type_ select-distinct)
> What is the motivation for changing the default?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

sinistral committed:
Test case fixes needed for modified rule syntax.

Because it no longer be assumed that all SPARQL functions and operators
occur in a `filter` block, `filter` must now be explicitly specified.

The change of default select-type default has been reverted, as this
breaks the standard query pattern, and is in fact not needed
(cf. https://github.com/UCDenver-ccp/kr/pull/4).

sinistral committed:
Test cases for new SPARQL operators.

sinistral committed:
Add support for operator validation.

It may be desirable in some instances to pre-validate the arguments that
are passed to an operator, rather than only relying on the backend
parsing of the generated SPARQL.

One example of this arises from how operators are implemented across
different backends.  For example, the IRI function across the various
platforms does not consistent handle language-tagged strings; some (like
Sesame) will do the conversion, while others (like Allegrograph) will
not.

This also provides a hook to insert validatation to help users with the
not-quite-SPARQL pattern syntax.

sinistral committed:
Remove defuct SPARQL body parse clauses.

A collection of SPARQL keywords were added both to the body parsing
condition, and to the `sparql-operators` map.  The latter actually takes
precendence; thus the clauses in the condition would never have been
hit.

sinistral commented:
I've pushed some clean up commits into this pull request.  Please review and let me know if you have any comment, but **do not merge** until we've had a chance to run the Reactome rules against these changes.
